src/css/index.ts

```diff
@@ -1,83 +1,186 @@
-import type { TW, CSSRules, CSSAtKeyframes, InlineDirective, Context, LazyInjected } from '../types'
+import type {
+  TW,
+  CSSRules,
+  CSSAtKeyframes,
+  InlineDirective,
+  LazyInjected,
+  Context,
+  CSSProperties,
+  Falsy,
+  MaybeThunk,
+} from '../types'
 
 import { tw as defaultTW, hash } from '../index'
+import * as is from '../internal/is'
+import { evalThunk, merge } from '../twind/util'
 
 export interface CSSDirective extends LazyInjected {
-  (context: { tw: TW }): CSSRules
+  (context: Context): CSSRules
 }
 
-const create = <P, R>(
-  factory: (parameter: P, key: string) => InlineDirective,
-): ((parameter: P) => R) => {
-  type Instance = (parameter: P) => R
+export type MaybeArray<T> = T | readonly T[]
 
-  // Cache css directive by their JSON stringified value
-  const instances = new WeakMap<TW, Instance>()
+export interface CSSFactory<T, I, R> {
+  (
+    strings: TemplateStringsArray,
+    ...interpolations: readonly MaybeThunk<MaybeArray<I | string | number | Falsy>>[]
+  ): R
+  (tokens: MaybeThunk<MaybeArray<T | Falsy>>): R
+  (...tokens: readonly MaybeThunk<T | Falsy>[]): R
+}
 
-  const build = (tw: TW): Instance => {
-    // Allows to invoke toString() on a directive:
-    // document.body.className = css({ color: 'white' })
-    function toString(this: InlineDirective): string {
-      return tw(this)
-    }
+function evaluateFunctions(this: TW, key: string, value: unknown): unknown {
+  return is.function(value) ? this(value as InlineDirective) : value
+}
 
-    const cache = new Map<string, R>()
-
-    const evaluateFunctions = (key: string, value: unknown): unknown =>
-      typeof value === 'function' ? tw(value as InlineDirective) : value
-
-    return (rules: P): R => {
-      const key = JSON.stringify(rules, evaluateFunctions)
-
-      let directive = cache.get(key)
-
-      if (!directive) {
-        directive = Object.defineProperties(factory(rules, key), {
-          valueOf: {
-            value: toString,
-          },
-          toString: {
-            value: toString,
-          },
-          // Allow twind to generate a unique id for this inline directive
-          // twind uses JSON.stringify which returns undefined for functions like this directive
-          // providing a toJSON function allows to include this directive in the id generation
-          toJSON: {
-            value: () => rules,
-          },
-        }) as R
-
-        cache.set(key, directive)
-
-        // Ensure the cache does not grow unlimited
-        if (cache.size > 10000) {
-          cache.delete(cache.keys().next().value)
-        }
+// eslint-disable-next-line @typescript-eslint/ban-types
+const lazy = <T extends Function>(
```

Comment:
This is a good lint rule; if you _can_ avoid `Function`, you should
```suggestion
const lazy = <T extends (...args: any[]) => any>(
```
I think that should work, but if there's something more specific you could put here, I'd go with that

Reply:
Thanks. See 69f59bf1